RGW:Multisite: Verify if the synced object is identical to source

prasakri opened the pull request:
Introduce an option 'rgw_copy_verify_object' which allows the object
copied from remote cluster through multisite sync is identical to the
source object. This is done by generating the MD5 checksum of the data
being copied and compared to the ETAG stored as part of the object's
attribute.

Signed-off-by: Prasad Krishnan <prasad.krishnan@flipkart.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

mykaul commented:
Please don't use MD5. Even if you don't use it for security, security scanners will flag it as a risk.
It's also not very fast. Just for hashing, you could use xxhash or others ( sha-256/512, or blake2b)

prasakri commented:
> Please don't use MD5. Even if you don't use it for security, security scanners will flag it as a risk.

The choice of hashing algorithm is determined by the customer and defaults to MD5 [as per AWS S3 API](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTCommonResponseHeaders.html). You [can also find](https://github.com/ceph/ceph/blob/master/src/rgw/rgw_tools.cc#L525) that every PUT object call invokes MD5 algorithm by default to calculate the ETAG.

mattbenjamin commented:
AWS S3 doesn't leave Etag computation to the client, it's assumed to be one of two computations (both using MD5) for PUT and mpu objects, respectively.  That said, because Etag is already MD5-based, I think this is a credible use.  That said, we're also introducing a checksum extension here: https://github.com/ceph/ceph/pull/30606 (thought to be complete except for mpu).

cbodley commented:
thanks for joining the standup today to discuss this!

one issue that came up is multipart uploads, which use a different algorithm for etag calculations. the sync client doesn't have enough information (the individual part sizes) to reproduce that calculation, but we could potentially get that from the original object's manifest

the reason that multipart uploads use a special strategy for etags is that it avoids having to process the entire data stream at once. but the GET requests we use to sync these objects in multisite do have to read the entire object, so we could do a separate checksum calculation in `RGWGetObj` as we stream out the response body. in this case, we don't have a final checksum until we've finished writing out the object data, so we can't provide that checksum in a header - instead, we'd have to encode it after the object data in the response body. this would be an extension to the s3 protocol, but sync already relies on one extension in the form of a `x-rgw-prepend-metadata` header that requests the object's attributes be encoded at the beginning of the response body. see `RGWRadosPutObj::handle_data()` and `process_attrs()` in rgw_rados.cc for the implementation

one major difference between the two approaches is the scope of integrity checking - the latter only protects against corruption over the sync network (and could presumably use TLS for integrity), while the former also covers the data at rest (albeit with a weak md5 hash)

my understanding of ETag in http is that it's intended to be used for caching and invalidation, rather than for data integrity. so if we're designing an integrity feature, i think we're better off choosing a more suitable hash than MD5, and could build on Matt's work in https://github.com/ceph/ceph/pull/30606

prasakri commented:
Hi @cbodley,

Thank you for the suggestions. Some additional thoughts:

While I agree with all the points mentioned by you in the comment above, I'm a little wary of introducing a protocol that appends the checksum at the end of the message for a couple of reasons.

1. While we have ready constructs to send a header (with custom fields) - both at the sender (`rgw::io::RestfulClient->send_header()`) and receiver (`RGWHTTPClient::receive_http_header()`), we'll have to make elaborate changes to append a tail with the checksum.
2. Alternatively, we could choose to embed the length of the checksum in the header and send the actual checksum as part of the message body with some additional calculation to track the extra data length alongside what's being done already in `RGWRadosPutObj::process_attrs()`.

However, I see additional risks to this approach (2) due to possibilities of data corruption as we're mutating the payload. Moreover, this method cannot detect the integrity of data at rest.

I was thinking if it makes sense to introduce a new message that specifically requests for the checksum of an object (along-with the algorithm to be used), and the receiver responds with the same after reading over the object once. For MPU objects, the checksum calculation can also involve additional verification of the ETag of the individual parts apart from calculating the checksum of the whole object to ensure data-at-rest integrity.

While this method entails obvious costs such as reading the entire object twice + additional REST API invocations, it appears to be safe and a simple approach.

Please feel free to pour in your thoughts on the same.

cbodley commented:
> While this method entails obvious costs such as reading the entire object twice + additional REST API invocations, it appears to be safe and a simple approach.

i don't like this approach, both because of the cost it incurs on every object sync, and because of the races you'd have to deal with. splitting one http request into two means that the source object could change in between, so the object you fetch could be different from the one you checksummed

note also that this doesn't really give you integrity of the data at rest, it only covers the time after recalculating the checksum

> the sync client doesn't have enough information (the individual part sizes) to reproduce that calculation, but we could potentially get that from the original object's manifest

can we explore this instead? `fetch_remote_obj()` does appear to get a copy of the manifest, but we filter it out in `RGWRadosPutObj::process_attrs()`:
```c++
src_attrs.erase(RGW_ATTR_MANIFEST); // not interested in original object layout
```
if we can use that manifest to discover the size of each part (if it is a multipart upload), we can reproduce the original etag calculation and use that to check integrity

prasakri commented:
@cbodley and I were discussing the hurdles to implementing the above-mentioned approach of using `RGW_ATTR_MANIFEST` to re-compute the ETag at the destination.

Basically, `RGWCompleteMultipart::execute()` takes individual Etag of the MPU parts and computes MD5 over it and appends the Etag string with the number of MPU parts.

However, we cannot repeat the process at the destination cluster using `RGWObjManifest` since it does not contain enough information to identify the individual MPU parts. The manifest contains information about the size of each RADOS object in the source cluster and not the size of the MPU parts. For e.g. if the MPU Upload size was 5MB and the head size was 4MB, a 10MB file contains manifest that shows the sizes as 4MB (head, part - 1) + 1MB (tail, part - 1) + 4MB (head, part - 2)+ 1MB (tail, part - 2).

Two thoughts were discussed as a solution for this (both of which have their short-comings):

1. Use the individual object names to identify the head and tail of the objects. It's not clear to me at the moment if this idea is entirely feasible with the existing information (since we must provide pool-name information to obtain raw object's name). Also, this will introduce a dependency on the naming policy and care must be taken to update the algorithm if the nomenclature undergoes a change.

2. Augment the `class RGWObjManifest` with additional information to identify MPU parts which will be updated during `RGWCompleteMultipart`. This can be done by incrementing the `struct_v` (which is currently 7), at the cost of losing backward compatibility.

@cbodley suggested that @mattbenjamin might have better suggestions or an opinion on any of those discussed above.

cbodley commented:
it looks like `RGWObjManifest::obj_iterator` does correctly track a `cur_part_id`, it just isn't publicly exposed. i added some debug logging for that in `operator++`, and downloaded a 128m multipart object split into 9 parts: https://gist.github.com/cbodley/ada02ee0749a34dec6366bc01b1d0fc5

prasakri commented:
> it looks like `RGWObjManifest::obj_iterator` does correctly track a `cur_part_id`, it just isn't publicly exposed. i added some debug logging for that in `operator++`, and downloaded a 128m multipart object split into 9 parts: https://gist.github.com/cbodley/ada02ee0749a34dec6366bc01b1d0fc5

@cbodley - yes, that certainly worked and I missed these private members.
I've updated the PR with another commit which adds support for MPU objects as well.
Kindly review the same.

cbodley commented on src/rgw/rgw_rados.cc:
```diff
@@ -3364,6 +3366,8 @@ class RGWRadosPutObj : public RGWHTTPStreamRWRequest::ReceiveCB
     const uint64_t lofs = data_len;
     data_len += size;
+    if (cct->_conf->rgw_copy_verify_object)
+      full_obj.append(bl);
```
it's not reasonable to buffer the entire object for this

once we finish reading the attributes, `process_attrs()` should decode the manifest and decide which etag strategy (multipart vs. non) to use. if it's a multipart manifest, we can store the array of part sizes

for the actual etag calculation, i think an extra filter (we `rgw::putobj::DataProcessor`) would be a good way to implement that. you could derive both a `MultipartEtagFilter` and `EtagFilter` that does the streaming md5 calculation on the data passed into `process()`. `process_attrs()` would attach one of those filters before the compression step

prasakri replied:
@cbodley - I agree that storing the entire object isn't quite efficient.

I've [re-written the patch](https://github.com/prasakri/ceph/commit/2677c4b88806d4af6d525157e7006c1b0ca1b964) to make the ETag verification as a filter based on your suggestion. Kindly review the same.

cbodley commented on src/rgw/rgw_rados.cc:
```diff
@@ -3318,6 +3322,58 @@ class RGWRadosPutObj : public RGWHTTPStreamRWRequest::ReceiveCB
+    /*
+     * Presently we don't support ETag based verification if compression or
+     * encryption is requested. We can enable simultaneous support once we have
+     * a mechanism to know the sequence in which the filters must be applied.
```
this should work fine with compression. as long as you compose this on top of the compression filter, the etag filter will see the uncompressed data

prasakri replied:
I wasn't sure about the sequence of filter invocation and how to ensure a certain order to it. How can I ensure that etag filter will be invoked only after compression filter?

cbodley replied:
you want it to be invoked _before_ compression so that you're calculating checksums on the uncompressed data

the ordering is determined by the pipelining of these filters; so you want your ETagVerifier's `next` filter to be the compression one. but you've already accomplished this, because you're initializing the ETagVerifier after the compression filter

prasakri replied:
Okay, will remove the check for `!plugin`

cbodley commented on src/rgw/rgw_rados.cc:
```diff
@@ -3290,7 +3293,8 @@ class RGWRadosPutObj : public RGWHTTPStreamRWRequest::ReceiveCB
       src_attrs.erase(RGW_ATTR_COMPRESSION);
-      src_attrs.erase(RGW_ATTR_MANIFEST); // not interested in original object layout
+      /* We need the manifest to recompute the ETag for verification */
+      manifest_bl = src_attrs[RGW_ATTR_MANIFEST];
```
we should still erase RGW_ATTR_MANIFEST from src_attrs, because fetch_remote_obj() writes those directly to the new head object:
```c++
      auto iter = src_attrs.find(RGW_ATTR_MANIFEST);
      if (iter != src_attrs.end()) {
        manifest_bl = std::move(iter->second);
        src_attrs.erase(iter);
      }
```

cbodley commented on src/rgw/rgw_rados.cc:
```diff
@@ -3946,6 +4012,22 @@ int RGWRados::fetch_remote_obj(RGWObjectCtx& obj_ctx,
+      if (cb.get_calculated_etag().compare(trimmed_etag)) {
+        ret = -EIO;
+        ldout(cct, 0) << "ERROR: source and destination objects don't match. Expected etag:"
+          << trimmed_etag << " Computed etag:" << cb.get_calculated_etag() << dendl;
```
we need to do this check before the call to `processor.complete()` - that's the transaction that makes the object visible

cbodley commented on src/rgw/rgw_rados.cc:
```diff
@@ -3384,6 +3440,16 @@ class RGWRadosPutObj : public RGWHTTPStreamRWRequest::ReceiveCB
+      etag_verifier_mpu->calculate_etag();
+      return etag_verifier_mpu->get_calculated_etag();
+    } else {
+      etag_verifier_atomic->calculate_etag();
```
this will crash if it's called when rgw_copy_verify_object==false. how about:
```diff
-   } else {
+   } else if (etag_verifier_atomic) {
      etag_verifier_atomic->calculate_etag();
      return etag_verifier_atomic->get_calculated_etag();
+   } else {
+     return "";
    }
```

cbodley commented on src/common/options.cc:
```diff
@@ -6287,6 +6287,13 @@ std::vector<Option> get_rgw_options() {
+    .set_description("Verify if the object copied is identical to its source")
+    .set_long_description(
+        "If true, this option computes the MD5 checksum of the data which is written at the"
+	"destination and checks if it is identical to the ETAG stored in the source."),
```
can we find a better name for this? i worry that 'copy' will be confused with the CopyObj operation. how about something like `rgw_fetch_remote_obj_integrity` or `rgw_sync_obj_integrity`? the description should mention that it applies to objects fetched from a remote over http, including multisite sync

cbodley commented on src/rgw/rgw_etag_verifier.h:
```diff
@@ -0,0 +1,68 @@
+
+  virtual void calculate_etag() = 0;
+  string get_calculated_etag() { return calculated_etag;}
+  virtual void append_part_ofs(uint64_t ofs) = 0;
```
this function is only called directly on `RGWPutObj_ETagVerifier_MPU`, so it doesn't need to be in the base class

prasakri commented on src/rgw/rgw_rados.cc:
```diff
@@ -3318,6 +3322,58 @@ class RGWRadosPutObj : public RGWHTTPStreamRWRequest::ReceiveCB
+      bool found = manifest.get_rule(0, &rule);
+      if (!found) {
+        lderr(cct) << "ERROR: manifest->get_rule() could not find rule" << dendl;
+        return -EIO;
```
I think it is fine to abort the download as a bad manifest indicates that the object header is corrupt. So, it is pointless to proceed further and download the rest of the object.

prasakri commented:
Hi @cbodley,
I've uploaded another patch to this PR which addresses the review comments. I wasn't sure if I should squash the review comments into the previous commit or create a new one (and chose the latter to give it the benefit of doubt). Let me know if you want the commits to be presented as a single one.

mattbenjamin commented on src/common/legacy_config_opts.h:
```diff
@@ -1434,7 +1434,7 @@ OPTION(rgw_curl_low_speed_limit, OPT_INT) // low speed limit for certain curl ca
 OPTION(rgw_copy_obj_progress, OPT_BOOL) // should dump progress during long copy operations?
 OPTION(rgw_copy_obj_progress_every_bytes, OPT_INT) // min bytes between copy progress output
-OPTION(rgw_copy_verify_object, OPT_BOOL) // verify if the copied object is identical to source
+OPTION(rgw_sync_obj_integrity, OPT_BOOL) // verify if the copied object from remote is identical to source
```
as in earlier comment by casey, let's not use the word "integrity" for etag comparison

prasakri replied:
@mattbenjamin - Does `rgw_sync_obj_etag_verify` sound better?
(BTW, `rgw_sync_obj_integrity` was chosen from one of @cbodley's suggestion)

mattbenjamin replied:
Sure, but yes:  I think rgw_sync_obj_etag_verify is ideal.

prasakri replied:
Done. Added another commit.

cbodley commented on src/rgw/rgw_rados.cc:
```diff
@@ -3384,6 +3442,19 @@ class RGWRadosPutObj : public RGWHTTPStreamRWRequest::ReceiveCB
   }
+
+  string get_calculated_etag() {
+    if (!cct->_conf->rgw_sync_obj_etag_verify)
```
the value of `rgw_sync_obj_etag_verify` might change at runtime, so we should only check its value once in `RGWRadosPutObj::process_attrs()`. later checks should be based on state that's local to RGWRadosPutObj

cbodley commented:
i'd like to see this tested in our multisite suite. can you please add `rgw sync obj etag verify: true` to `qa/suites/rgw/multisite/overrides.yaml`?

cbodley commented on src/rgw/rgw_rados.cc:
```diff
@@ -3384,6 +3444,18 @@ class RGWRadosPutObj : public RGWHTTPStreamRWRequest::ReceiveCB
+
+  string get_calculated_etag() {
+    if (obj_type == OBJ_TYPE_ATOMIC) {
+      etag_verifier_atomic->calculate_etag();
```
can just do if (etag_verifier_atomic) ... else if (etag_verifier_mpu) without the extra enum

prasakri replied:
But I see that there's a third case as well where we have to return the empty string. I used to `OBJ_TYPE_UNINIT` to track the same.

cbodley commented on src/rgw/rgw_rados.cc:
```diff
@@ -3937,6 +4009,21 @@ int RGWRados::fetch_remote_obj(RGWObjectCtx& obj_ctx,
     set_mtime_weight.init(set_mtime, svc.zone->get_zone_short_id(), pg_ver);
   }
 
+  if (cct->_conf->rgw_sync_obj_etag_verify) {
```
as above, this shouldn't depend on rgw_sync_obj_etag_verify either
```c++
std::string calculated_etag = cb.get_calculated_etag();
if (!calculated_etag.empty()) {
```

prasakri replied:
I think I will fallback on `obj_type` member in `class RGWRadosPutObj` to find out if we need to perform etag verification.

ivancich commented:
QA results:
- http://pulpito.ceph.com/ivancich-2020-06-08_20:10:46-rgw-wip-eric-testing-1-distro-basic-smithi/
- http://pulpito.ceph.com/ivancich-2020-06-08_20:11:05-rgw-wip-eric-testing-1-distro-basic-smithi/

cbodley commented:
it looks like this PR is the root cause of https://tracker.ceph.com/issues/45992. the same test_multipart_object_sync() failures were present in the posted QA results

cbodley commented:
it looks like the issue is related to compression. we're reading the source object's manifest to infer the part sizes, then using those part sizes to calculate checksums on the corresponding ranges of object data. but when the source object is compressed, its manifest contains the compressed part sizes, so we end up checksumming the wrong ranges and incorrectly report that `source and destination objects don't match`

it looks like there's logic in `RGWGetObj_Decompress::fixup_range()` to convert these compressed offsets into logical offsets, so i'll try to get that working; otherwise we may need to disable verification of compressed objects the same way we do with encrypted ones

ivancich commented:
Tremendous progress, @cbodley!

cbodley commented:
@prasakri if you're available, i'd appreciate your help in reviewing https://github.com/ceph/ceph/pull/37238

prasakri commented:
> @prasakri if you're available, i'd appreciate your help in reviewing #37238

@cbodley - Apologies for the late reply and for the regression. I did not see this notification (will check github settings to see if there's a way to receive email upon mentions). Glad that you fixed it.

yanghonggang commented:
@prasakri @cbodley 
Thank you for your contribution. Is there any plan to backport #33492 and #37238 to nautilus?
I add a tracker for #33492, but failed  to create a backport tracker.

```text
$ ./src/script/backport-create-issue --user yanghonggang --password XXXXX --dry-run  48798
INFO:root:Dry run: nothing will be written to Redmine
INFO:root:Redmine username and password were provided; using them
INFO:root:Processing issue list ->48798<-
INFO:root:Processing 0 issues with status Pending Backport
INFO:root:Processed 0 issues         
```

prasakri commented:
> Is there any plan to backport #33492 and #37238 to nautilus?
@yanghonggang - Not in my plan, although I would assume that patching it will work seamlessly given that there are few changes between nautilus and master in these parts of the code.

yanghonggang commented:
> > Is there any plan to backport #33492 and #37238 to nautilus?
> > @yanghonggang - Not in my plan, although I would assume that patching it will work seamlessly given that there are few changes between nautilus and master in these parts of the code.

Thank you for your response. I will backport #33492 and #37238 to nautilus.

cbodley commented:
@yanghonggang can you please prepare the octopus backport as well? we'll need to make sure that users upgrading from nautilus to octopus don't lose features

yanghonggang commented:
> @yanghonggang can you please prepare the octopus backport as well? we'll need to make sure that users upgrading from nautilus to octopus don't lose features

OK
